experimental/alp/lib/Transforms/extract_kernel_pass.cpp

```diff
@@ -0,0 +1,186 @@
+  });
+
+  llvm::SmallVector<Type, 1> newtypes;
+  // We are not done yet. We need to merge the block into the entry block. To do this:
+  // 1 If an operation in the block is using a value coming from the block
+  //   argument, add the value as function argument and replace the value with it
+  // 2 If an operation in the block is using a value generated outside the
+  //   block, simply replace its value with a funciton argument
+
+  // Step 1: get all the block arguments, add them as function arguments and
+  // replece their use inside the block
+  for (auto block_arg : block->getArguments())
+  {
+    func_op.insertArgument(vals.size(), block_arg.getType(), {});
+    auto arg = func_op.getArgument(vals.size());
+    block_arg.replaceAllUsesWith(arg);
+    newtypes.push_back(block_arg.getType());
+  }
+
+  // Step 2: replace all the values that are pointing outside the block and
+  // replace them with function arguments
+  auto args = func_op.getArguments();
+  for (unsigned i = 0; i < vals.size(); i++)
+  {
+    auto val = vals[i];
+    auto arg = args[i];
+    val.replaceUsesWithIf(arg, [&](OpOperand &op)
+                          { return dom_info.dominates(block, op.getOwner()->getBlock()); });
+  }
+
+  // Save some information about the original block. Once the block is merged
+  // inside the entry block these information won't be available anymore
+  bool has_no_successor = block->hasNoSuccessors();
+  Block *succ = (has_no_successor ? nullptr : block->getSuccessor(0));
+
+  // Remove all arguments from the block signature
+  for (unsigned i = 0; i<block->getNumArguments(); i++){
+    block->eraseArgument(i);
+  }
+
+  // Merge block into entry_block (this destroys block)
+  rewriter.mergeBlocks(block, entry_block);
+
+  // Add a returnOp into the block to properly terminate it
+  rewriter.setInsertionPointToEnd(entry_block);
+  rewriter.create<ReturnOp>(loc);
+
+  // We are done with the callee. Now we have to work on the caller. The overall
+  // idea is to insert a new_block right before the successor of the old block.
+  // If the old block has no successors, then add it at the end of the region
+  Block *new_block = nullptr;
+  if (has_no_successor)
+  {
+    new_block = rewriter.createBlock(region, region->end(), newtypes);
+  }
+  else
+  {
+    new_block =rewriter.createBlock(succ, newtypes);
+  }
+
+  // Remember to add the block arguments as inputs to the function
+  for (unsigned i = 0; i<new_block->getNumArguments(); i++){
+    vals.push_back(new_block->getArgument(i));
+  }
+
+  // Create the call 
+  rewriter.create<CallOp>(loc, func_op, vals);
+
+  if (add_yield){
+    rewriter.create<scf::YieldOp>(loc);
+  }
+}
+
+
+// Walk the for loops and find the one that as operands. In GEMM is the micro-kernel.
+// TODO: we should have the linalg::split to signal the microkernel of the operation
+// and use it to run the function extractor if needed
+struct ExtractKernelPass : public ExtractKernelPassBase<ExtractKernelPass> {
```

Comment:
I am unclear why this is needed per-se.
If your objective is to improve the level of control and try to scale this up, be aware that we are exploring a few different things that seem to not intersect with this pass.

Reply:
It's unclear to me as well :) The only thing I know is that if we don't extract the kernel in a separate function there is significant spilling. This does not happen always, but for specific cases where we particularly stress the scheduler (e.g., 12x8 inner  tiling). We are investigating the LLVM scheduler, but this for now seemed the only ( yet hacky) solution

Reply:
Ah ok, good to know, thanks for sharing. I have seen similar behavior myself in the past but not on these examples here. Makes sense to me, I'll think about it more.

Could you maybe add this rationale to the doc of you pass / in the cpp code somewhere ?